### ISSUE TITLE
F87 and ARIA 24 seem to contradict in the use of CSS content  property

### ISSUE THREAD

**Issue by mraccess77** [w3c/wcag#1554]
[ARIA24 example 3](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA24#example-3-email-icon-font-in-a-link-without-visible-text) (and maybe others) indicates the content property can be used while [F87](https://www.w3.org/WAI/WCAG21/Techniques/failures/F87.html) says it can't be relied upon.

**Comment by patrickhlauke** [comment]
In the case of F87, the `content` property was relied on to convey meaningful ("non-decorative") text content (i.e. exposed to screen readers etc), and it's saying that's a failure because it can't be relied on to be exposed consistently as content to users (though support across browser/AT has considerably improved since the time that failure technique was written, and as far as I know all modern browsers and AT combinations now expose CSS-generated content correctly).

ARIA24 on the other hand uses `content` to display a visual icon type character, but its actual accessible name (what gets passed on to screen readers) is given by `aria-label` in all cases, so even assuming that some browser/AT combos out there still don't expose CSS-generated content correctly, it doesn't matter because it's explicitly redefined here via `aria-label` anyway.

So, not a contradiction I'd say (though F87 could probably do with some more recent note/clarification about current browser/AT support; and perhaps a better/more holistic angle here would also be that if the CSS fails to load or something, then all of a sudden your page would be missing actual text content for ALL users - AT and non-AT users alike - and that CSS-generated content will not generally/currently be indexed by search engines etc...so not relying on `content` here has collateral issues, but from the browser/AT point of view it may nowadays actually be a solved issue)

**Comment by mraccess77** [comment]
The rub is the check in F87
If the inserted content is not decorative, check that the information is provided to assistive technologies and is also available when CSS is turned off.

**Comment by JAWS-test** [comment]
See https://github.com/w3c/wcag/issues/907

**Comment by alastc** [comment]
From memory, F87 was kept as folks from the low-vision task force were concerned that these types of icons / content disappear when fonts and/or CSS are over-ridden.

I do wonder if it should be converted to a failure technique based more on font-icons, as that seems to be the more common example?

A quick patch would be to get ARIA24 to use another CSS method than content (e.g. positioning of a background), but that's just a technicality.

**Comment by patrickhlauke** [comment]
I think, as @mraccess77 suggests, that the main problem with F87 is the check that essentially says "test if content disappear with custom/no CSS, and if so it fails 1.3.1". that goes, i'd argue, way beyond what 1.3.1 normatively requires. or rather, it lacks nuance, as the same meaning/content may well be conveyed otherwise (with ARIA, or additional visually hidden text, etc), which does pass 1.3.1 / fulfill the requirement of being *programmatically determinable*. Yes, these latter ones won't help sighted users that also change/disable styles, but that's not the point of 1.3.1 per se. also, keeping in mind that even using pure CSS generated content does, in my view, pass 1.3.1, as the content is exposed correctly in the accessibility tree in modern browsers (something that didn't use to be the case years ago). if CSS is relied on, then this seems perfectly kosher to me.

as such, I don't think F87 is a valid failure technique (in the sense that this always fails 1.3.1). rather, it's more of a best practice that goes beyond the normative ask of 1.3.1?

**Comment by patrickhlauke** [comment]
the issue with reader mode there is that CSS is relied on as an accessibility supported technology, but then reader mode simply drops it on the floor. arguably a failing/shortcoming of reader mode. question then is does the fact that reader mode does away with this effectively invalidate all reliance on CSS as accessibility supported technology, or is reader mode essentially an outlier and it can be said that it's fine to rely on CSS because users have technologies readily available at their disposal (i.e. NOT enabling reader mode) to still access the content as authored.

**Comment by patrickhlauke** [comment]
as user styles can completely break an experience even if it doesn't rely on CSS (just a silly `* { display: none; }` is all that's needed), they're generally not taken into account when evaluating content against most SCs. and if a site chooses to use CSS as technology that's relied upon for conformance (which is generally the rule these days, as beyond just generated content there's other things like showing/hiding content, modal dialogs, etc that would simply not work if CSS is ignored/overridden incorrectly) then i'd find it hard to fail it normatively.

ARIA24 could do with being tweaked/updated (as mentioned earlier in the thread)  as yes, it seems to enshrine the idea that user styles DO need to be taken into account. and F87 is now a bit long in the tooth and hasn't really kept up pace with browser changes since 2008... (as is the case with many 2.0 techniques)

**Comment by awkawk** [comment]
This came up in a discussion with @jnurthen today and we think that this failure is not valid and should be removed. Want me to make a PR to remove it? :)

**Comment by detlevhfischer** [comment]
+1 for deletion
